scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/SchedulingMethodImpl.java

```diff
@@ -187,28 +188,57 @@ public int schedule() {
             currentPolicy.setRMState(rmState);
             Set<String> freeResources = rmState.getFreeNodes();
             if (logger.isDebugEnabled()) {
-                logger.debug("eligible nodes : " + freeResources);
+                logger.debug("eligible nodes : " + (freeResources.size() < 5 ? freeResources : freeResources.size()));
             }
-            //if there is no free resources, stop it right now
+            //if there is no free resources, stop it right now without starting any task
             if (freeResources.isEmpty()) {
-                return numberOfTaskStarted;
+                return 0;
             }
 
             // ask the policy all the tasks to be schedule according to the jobs list.
+            LinkedList<EligibleTaskDescriptor> fullListOfTaskRetrievedFromPolicy = currentPolicy.getOrderedTasks(descriptors);
+
+            //if there is no task to scheduled, return without starting any task
+            if (fullListOfTaskRetrievedFromPolicy == null || fullListOfTaskRetrievedFromPolicy.isEmpty()) {
+                return 0;
+            }
 
-            LinkedList<EligibleTaskDescriptor> taskRetrievedFromPolicy = currentPolicy.getOrderedTasks(descriptors);
+            schedulingService.unlockJobsToSchedule(toUnlock.values());
+            toUnlock = null;
+
+            int numberOfTaskStarted = selectAndStartTasks(currentPolicy,
+                                                          jobMap,
+                                                          freeResources,
+                                                          fullListOfTaskRetrievedFromPolicy);
 
-            //if there is no task to scheduled, return
-            if (taskRetrievedFromPolicy == null || taskRetrievedFromPolicy.isEmpty()) {
+            if (freeResources.isEmpty()) {
+                return numberOfTaskStarted;
+            }
+            if (activeObjectCreationRetryTimeNumber == 0) {
                 return numberOfTaskStarted;
             }
 
-            if (logger.isDebugEnabled()) {
-                logger.debug("eligible tasks : " + taskRetrievedFromPolicy);
+            return numberOfTaskStarted;
+        } finally {
+            if (toUnlock != null) {
+                schedulingService.unlockJobsToSchedule(toUnlock.values());
             }
+        }
+    }
 
-            schedulingService.unlockJobsToSchedule(toUnlock.values());
-            toUnlock = null;
+    private int selectAndStartTasks(Policy currentPolicy, Map<JobId, JobDescriptor> jobMap, Set<String> freeResources,
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Refactor this method to reduce its Cognitive Complexity from 69 to the 15 allowed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3776)